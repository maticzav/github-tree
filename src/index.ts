export { loadTreeFromPath } from './fs'
export { commit, CommitInput } from './github'
export { utf8, base64, encoding, content, mapFiles, mapPaths } from './tree'
