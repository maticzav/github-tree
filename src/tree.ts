import * as path from 'path'

import { Dict, not } from './utils'

/**
 * Represents a Github file/folder structure.
 */
export type Tree = { [path: string]: File }

export type File = { content: string; encoding: 'utf-8' | 'base64' }

/* Accessors, Getters */

/**
 * Returns the content of a file.
 * @param file
 */
export function content(file: File): string {
  return file.content
}

/**
 * Returns the encoding of a file.
 * @param file
 */
export function encoding(file: File): File['encoding'] {
  return file.encoding
}

/**
 * Creates an utf-8 encoded file.
 *
 * @param content
 */
export function utf8(content: string): File {
  return { content, encoding: 'utf-8' }
}

/**
 * Creates a base64 encoded file.
 *
 * @param content
 */
export function base64(content: string): File {
  return { content, encoding: 'base64' }
}

/**
 * Returns the files that are not nested.
 * @param tree
 */
export function getTreeFiles(tree: Tree): Dict<File> {
  return Object.fromEntries(
    Object.keys(tree)
      .filter(isFileInThisFolder)
      .map(name => [name, tree[name]]),
  )
}

/**
 * Returns a dictionary of remaining subtrees.
 * @param tree
 */
export function getTreeSubTrees(tree: Tree): Dict<Tree> {
  return Object.keys(tree)
    .filter(not(isFileInThisFolder))
    .reduce<Dict<Tree>>((acc, filepath) => {
      const [subTree, newFilepath] = shiftPath(filepath)
      if (!acc.hasOwnProperty(subTree)) {
        acc[subTree] = {}
      }
      acc[subTree][newFilepath] = tree[filepath]
      return acc
    }, {})
}

/**
 * Shifts path by one.
 * Returns the shifted part as first argument and remaining part as second.
 *
 * @param filepath
 */
function shiftPath(filepath: string): [string, string] {
  const [dir, ...dirs] = filepath.split('/').filter(Boolean)
  return [dir, dirs.join('/')]
}

/**
 * Determines whether a path references a direct file
 * or a file in the nested folder.
 *
 * "/src/index.ts" -> false
 * "/index.ts" -> true
 * "index.ts" -> true
 *
 * @param filePath
 */
function isFileInThisFolder(filePath: string): boolean {
  return ['.', '/'].includes(path.dirname(filePath))
}
