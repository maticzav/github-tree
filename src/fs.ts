import * as fs from 'fs'
import * as path from 'path'
import { mapKeys } from './utils'
import { Tree, utf8 } from './tree'

/**
 * Loads a tree of utf-8 decoded files at paths.
 *
 * @param path
 */
export function loadTreeFromPath(
  root: string,
  ignore: (string | RegExp)[],
): Tree {
  const filePaths = fs.readdirSync(root, { encoding: 'utf-8' })
  const tree = filePaths
    .filter(filePath => !ignore.some(glob => RegExp(glob).test(filePath)))
    .flatMap(filePath => {
      /**
       * Calculates the absolute file path from infomation about where the root is.
       */
      const rootFilePath = path.resolve(root, filePath)
      if (fs.lstatSync(rootFilePath).isDirectory()) {
        return Object.entries(
          mapKeys(loadTreeFromPath(rootFilePath, ignore), key =>
            unshift(filePath, key),
          ),
        )
      } else {
        const file = utf8(fs.readFileSync(rootFilePath, { encoding: 'utf-8' }))
        return [[filePath, file]]
      }
    })

  return Object.fromEntries(tree)
}

/**
 * Adds a folder to the path.
 * @param pre
 * @param path
 */
function unshift(pre: string, path: string): string {
  return [pre, ...path.split('/').filter(Boolean)].join('/')
}
