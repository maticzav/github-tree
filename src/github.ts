import { Octokit } from '@octokit/rest'

import { Tree, File, getTreeFiles, getTreeSubTrees } from './tree'
import { Dict, mapEntriesAsync, not } from './utils'

/**
 * Input variables for commit method.
 */
export type CommitInput = {
  owner: string
  repo: string
  message: string
  ref: string
  tree: Tree
}

/**
 * Create a commit to Github using Github API v3.
 *
 * @param github
 * @param CommitInput
 */
export async function commit<O extends Octokit>(
  github: O,
  { owner, repo, message, ref, tree }: CommitInput,
): Promise<Octokit.GitCreateCommitResponse> {
  /**
   * Variables prefixed with g reference values that exist on Github.
   */
  const gTree = await createGhTree(github, { owner, repo }, tree)
  const gRef = await github.git
    .getRef({ owner, repo, ref })
    .then(res => res.data)

  const commit = await github.git
    .createCommit({
      owner,
      repo,
      message,
      tree: gTree.sha,
      parents: [gRef.object.sha],
    })
    .then(res => res.data)

  await github.git
    .updateRef({
      owner,
      repo,
      ref,
      sha: commit.sha,
    })
    .then(res => res.data)

  return commit
}

/**
 * Recursively creates a tree commit by creating blobs and generating
 * trees on folders.
 *
 * @param github
 * @param tree
 */
async function createGhTree(
  github: Octokit,
  { owner, repo }: { owner: string; repo: string },
  tree: Tree,
): Promise<{ sha: string }> {
  /**
   * Uploads blobs and creates subtrees.
   */
  const blobs = await mapEntriesAsync(getTreeFiles(tree), content =>
    createGhBlob(github, { owner, repo }, content),
  )
  const trees = await mapEntriesAsync(getTreeSubTrees(tree), subTree =>
    createGhTree(github, { owner, repo }, subTree),
  )

  return github.git
    .createTree({
      owner,
      repo,
      tree: [
        ...Object.entries(trees).map(([treePath, { sha }]) => ({
          mode: '040000' as const,
          type: 'tree' as const,
          path: treePath,
          sha,
        })),
        ...Object.entries(blobs).map(([filePath, { sha }]) => ({
          mode: '100644' as const,
          type: 'blob' as const,
          path: filePath,
          sha,
        })),
      ],
    })
    .then(res => res.data)
}

/**
 * Creates a Github Blob from a File.
 *
 * @param github
 * @param param1
 * @param file
 */
async function createGhBlob(
  github: Octokit,
  { owner, repo }: { owner: string; repo: string },
  file: File,
): Promise<Octokit.GitCreateBlobResponse> {
  return github.git
    .createBlob({ owner, repo, content: file.content, encoding: file.encoding })
    .then(res => res.data)
}
