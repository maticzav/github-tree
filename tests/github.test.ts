import { Octokit } from '@octokit/rest'
import * as nock from 'nock'
import * as path from 'path'

import { commit, loadTreeFromPath, CommitInput } from '../src'

describe('github:', () => {
  beforeAll(() => {
    nock.disableNetConnect()
  })

  afterAll(() => {
    nock.enableNetConnect()
  })

  test(
    'commit',
    async () => {
      expect.assertions(5)

      const params: CommitInput = {
        owner: 'maticzav',
        repo: 'github-tree',
        message: 'My test commit!',
        ref: 'heads/master',
        tree: loadTreeFromPath(
          path.resolve(__dirname, './__fixtures__/example'),
          [],
        ),
      }

      /* Mocks */

      let blobs: { [sha: number]: string } = {}

      const createBlobEndpoint = jest.fn().mockImplementation((uri, body) => {
        const sha = Object.keys(blobs).length
        blobs[sha] = body.content
        return { url: 'url', sha }
      })

      let trees: { [sha: number]: string } = {}

      const createTreeEndpoint = jest.fn().mockImplementation((uri, body) => {
        const sha = Object.keys(trees).length
        trees[sha] = body.tree
        return { sha: sha, url: 'url', tree: body.tree }
      })

      const parentSha = Math.floor(Math.random() * 1000).toString()

      const getRefEndpoint = jest.fn().mockImplementation((uri, body) => {
        return { object: { sha: parentSha } }
      })

      const commitSha = Math.floor(Math.random() * 1000).toString()

      const createCommitEndpoint = jest.fn().mockImplementation((uri, body) => {
        expect(body.parents).toEqual([parentSha])
        expect(body.message).toBe(params.message)
        return { sha: commitSha }
      })

      const updateRefEndpoint = jest.fn().mockImplementation((uri, body) => {
        expect(body.sha).toBe(commitSha)
        return { object: { sha: '' } }
      })

      /* Mocks */

      nock('https://api.github.com')
        .post('/repos/maticzav/github-tree/git/blobs')
        .reply(200, createBlobEndpoint)
        .persist()

      nock('https://api.github.com')
        .post('/repos/maticzav/github-tree/git/trees')
        .reply(200, createTreeEndpoint)
        .persist()

      nock('https://api.github.com')
        .get('/repos/maticzav/github-tree/git/ref/heads/master')
        .reply(200, getRefEndpoint)

      nock('https://api.github.com')
        .post('/repos/maticzav/github-tree/git/commits')
        .reply(200, createCommitEndpoint)

      nock('https://api.github.com')
        .patch('/repos/maticzav/github-tree/git/refs/heads/master')
        .reply(200, updateRefEndpoint)

      const octokit = new Octokit({})
      await commit(octokit, params)

      /* Tests */

      expect(blobs).toMatchSnapshot()
      expect(trees).toMatchSnapshot()
    },
    5 * 60 * 1000,
  )
})
