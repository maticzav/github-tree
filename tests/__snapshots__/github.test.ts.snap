// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github: commit 1`] = `
Object {
  "0": "# Template Fixture
",
  "1": "{
  \\"compilerOptions\\": {
    \\"target\\": \\"es2017\\",
    \\"module\\": \\"commonjs\\",
    \\"lib\\": [\\"dom\\", \\"es2017\\", \\"es2019\\"],
    \\"rootDir\\": \\"src\\",
    \\"outDir\\": \\"dist\\"
  }
}
",
  "2": "",
  "3": "function helloWorld() {
  return 'hi!'
}

helloWorld()
",
  "4": "export function nested() {
  return true
}
",
}
`;

exports[`github: commit 2`] = `
Object {
  "0": Array [
    Object {
      "mode": "100644",
      "path": "ignored",
      "sha": 2,
      "type": "blob",
    },
  ],
  "1": Array [
    Object {
      "mode": "100644",
      "path": "nested.ts",
      "sha": 4,
      "type": "blob",
    },
  ],
  "2": Array [
    Object {
      "mode": "040000",
      "path": "folder",
      "sha": 1,
      "type": "tree",
    },
    Object {
      "mode": "100644",
      "path": "index.ts",
      "sha": 3,
      "type": "blob",
    },
  ],
  "3": Array [
    Object {
      "mode": "040000",
      "path": "ignore",
      "sha": 0,
      "type": "tree",
    },
    Object {
      "mode": "040000",
      "path": "src",
      "sha": 2,
      "type": "tree",
    },
    Object {
      "mode": "100644",
      "path": "README.md",
      "sha": 0,
      "type": "blob",
    },
    Object {
      "mode": "100644",
      "path": "tsconfig.json",
      "sha": 1,
      "type": "blob",
    },
  ],
}
`;
